### PR TITLE
[DevEnv] Improve patch_remote script

### PR DESCRIPTION
**Pull request by adamdelman**
Improve script for faster building and support for pushing mlrun image

### CODE REVIEW

File: automation/patch_igz/patch_remote.py [hunk 1]
```diff
@@ -40,101 +40,120 @@ class Consts:
...
+    def patch(self):
+        version = self._get_current_version()
+        image_tag = self._get_image_tag(version)
+        targets = []
+        if self._patch_api:
+            targets.append("api")
+        if self._patch_mlrun_image:
+            targets.append("mlrun")
+        if self._patch_log_collector_image:
+            targets.append("log-collector")
```
Comment (TomerShor) [inline automation/patch_igz/patch_remote.py]: You can move the string to the `Consts` class as you use them in multiple places

File: automation/patch_igz/patch_remote.py [hunk 2]
```diff
@@ -40,101 +40,120 @@ class Consts:
...
+        if not targets:
+            raise ValueError("No targets to patch")
+
+        target_to_built_images = self._make_targets(
+            targets=targets,
+            image_tag=image_tag,
+        )
+        # Docker login and push images
         self._docker_login_if_configured()
```
Comment (TomerShor) [inline automation/patch_igz/patch_remote.py]: move this before making the targets to fail fast if needed

File: automation/patch_igz/patch_remote.py [hunk 3]
```diff
@@ -40,101 +40,120 @@ class Consts:
...
-            )
-            built_images.append(built_log_collector_image)
-
-        built_images = self._tag_images_for_multi_node_registries(built_images)
+        built_images = self._tag_images_for_multi_node_registries(
+            target_to_built_images.values()
+        )
         self._push_docker_images(built_images)
 
+        # Connect to the first node and start deployment process
```
Comment (TomerShor) [inline automation/patch_igz/patch_remote.py]:
```suggestion
        # Connect to the first node and start deployment patching process
```

File: automation/patch_igz/patch_remote.py [hunk 4]
```diff
@@ -40,101 +40,120 @@ class Consts:
...
         node = self._nodes[0]
         self._connect_to_node(node)
-        try:
-            self._replace_deploy_policy()
-            self._replace_deployment_images(self.Consts.api_container, built_api_image)
-            if self._patch_log_collector:
-                # sanity
-                if not built_log_collector_image:
-                    raise RuntimeError("Log collector image not built")
+        if self._patch_api:
```
Comment (TomerShor) [inline automation/patch_igz/patch_remote.py]: what if we only want to patch the log collector?
or the mlrun image? (not api)

File: automation/patch_igz/patch_remote.py [hunk 5]
```diff
@@ -40,101 +40,120 @@ class Consts:
...
+                    "docker",
+                    "login",
+                    docker_registry or "",
+                    "--username",
+                    registry_username,
+                    "--password",
+                    registry_password,
+                ],
+                live=True,
             )
```
Comment (TomerShor) [inline automation/patch_igz/patch_remote.py]: Why not failing on error?

File: automation/patch_igz/patch_remote.py [hunk 6]
```diff
@@ -224,12 +254,13 @@ def _push_docker_images(self, built_images):
...
+            ]
             self._exec_local(
-                [
-                    "docker",
-                    "push",
-                    image,
-                ],
+                cmd,
                 live=True,
             )
```
Comment (TomerShor) [inline automation/patch_igz/patch_remote.py]: Not that it's wrong, but is there a reasoning for this change?

File: docs/runtimes/dask-mlrun.ipynb
Comment (TomerShor) [inline docs/runtimes/dask-mlrun.ipynb:1]: How is this related to this PR?

File: automation/patch_igz/patch_remote.py [hunk 7]
```diff
@@ -590,13 +621,42 @@ def _resolve_overwrite_registry(self):
...
+    "--mlrun",
+    is_flag=True,
+    help="Deploy the mlrun image",
+)
+@click.option(
+    "-ap",
+    "--api",
+    is_flag=True,
+    help="Deploy the mlrun API image",
 )
```
Comment (TomerShor) [inline automation/patch_igz/patch_remote.py]: Thinking out loud: Maybe we should make this "True" by default, and change it to "skip-api", as most of the time we will use this script to patch the api

**Review by TomerShor — CHANGES_REQUESTED** [review]
Looking good!
Added a couple of comment below

File: automation/patch_igz/patch_remote.py [hunk 8]
```diff
@@ -40,101 +40,120 @@ class Consts:
...
+                [
+                    "docker",
+                    "login",
+                    docker_registry or "",
+                    "--username",
+                    registry_username,
+                    "--password",
+                    registry_password,
+                ],
+                live=True,
```
Comment (liranbg) [inline automation/patch_igz/patch_remote.py]: suggest to undo to my changes so allow reading password from stdin and not as plain text

File: automation/patch_igz/patch_remote.py (hunk 6 above)
Comment (adamdelman) [reply automation/patch_igz/patch_remote.py]: fixed

File: automation/patch_igz/patch_remote.py [hunk 9]
```diff
@@ -34,97 +33,123 @@
...
 
-            if self._reset_db:
-                self._reset_mlrun_db()
-            else:
-                self._rollout_deployment()
+                if self._patch_log_collector_image:
+                    self._replace_deployment_images(
+                        Consts.log_collector_container,
+                        target_to_built_images[Consts.log_collector],
+                    )
```
Comment (alonmr) [inline automation/patch_igz/patch_remote.py]: this should be outside of `if not self._skip_patch_api:` to support patching only the log collector

File: automation/patch_igz/patch_remote.py [hunk 10]
```diff
@@ -34,97 +33,123 @@
 coloredlogs.install(level=log_level, logger=logger, fmt=fmt)
 
 
-class MLRunPatcher:
-    class Consts:
-        mandatory_fields = {"DATA_NODES", "SSH_USER", "SSH_PASSWORD", "DOCKER_REGISTRY"}
-        api_container = "mlrun-api"
-        log_collector_container = "mlrun-log-collector"
+class Consts:
```
Comment (alonmr) [inline automation/patch_igz/patch_remote.py]:
```suggestion
class Constants:
```

File: automation/patch_igz/patch_remote.py [hunk 11]
```diff
@@ -34,97 +33,123 @@
...
-            self._replace_deploy_policy()
-            self._replace_deployment_images(self.Consts.api_container, built_api_image)
-            if self._patch_log_collector:
-                # sanity
-                if not built_log_collector_image:
-                    raise RuntimeError("Log collector image not built")
+        if not self._skip_patch_api:
+            try:
+                # Replace deployment policies and images
+                self._replace_deploy_policy()
```
Comment (alonmr) [inline automation/patch_igz/patch_remote.py]: please change the name it's not very indicative 
```suggestion
                self._patch_deployment_from_file()
```

**Review by alonmr — CHANGES_REQUESTED** [review]
Very much needed! Some small comments

File: automation/patch_igz/patch_remote.py [hunk 12]
```diff
@@ -34,97 +33,123 @@
...
-                "log-collector", image_tag, "log-collector"
-            )
-            built_images.append(built_log_collector_image)
-
-        built_images = self._tag_images_for_multi_node_registries(built_images)
+        target_to_built_images = self._make_targets(
+            targets=targets,
+            image_tag=image_tag,
+        )
+        # Docker login and push images
```
Comment (TomerShor) [inline automation/patch_igz/patch_remote.py]:
```suggestion
        # build and push images
```

File: automation/patch_igz/patch_remote.py (hunk 4 above)
Comment (TomerShor) [reply automation/patch_igz/patch_remote.py]: This comment still stands..

File: automation/patch_igz/patch_remote.py [hunk 13]
```diff
@@ -34,97 +33,122 @@
...
         logger.info(
-            "Deployed branch successfully! Yay! (Note this may not survive system restarts)"
+            "Deployed branch successfully! (Note: This may not survive system restarts)"
         )
 
     def _docker_login_if_configured(self):
         registry_username = self._config.get("REGISTRY_USERNAME")
         registry_password = self._config.get("REGISTRY_PASSWORD")
-        registry_url = self._config.get("REGISTRY_URL")
+        docker_registry = self._config.get("DOCKER_REGISTRY")
```
Comment (alonmr) [inline automation/patch_igz/patch_remote.py:145]: need to remove `REGISTRY_URL` from `automation/patch_igz/patch_env_template.yml`

**Review by alonmr — APPROVED** [review]
